elasticdl/python/data/odps_io.py

```diff
@@ -266,22 +269,41 @@ def _estimate_cache_batch_count(self, columns, table_size, batch_size):
 
         sample_size = 10
         max_cache_batch_count = 50
-        upper_bound = 20 * 1000000
 
         if table_size < sample_size:
             return 1
 
-        batch = self.read_batch(start=0, end=sample_size, columns=columns)
-
-        size_sample = _nested_list_size(batch)
-        size_per_batch = size_sample * batch_size / sample_size
+        record_size = self._estimate_record_size(columns)
+        size_per_batch = record_size * batch_size
 
         # `size_per_batch * cache_batch_count` will
         # not exceed upper bound but will always greater than 0
-        cache_batch_count_estimate = max(int(upper_bound / size_per_batch), 1)
+        cache_batch_count_estimate = max(
+            int(MAX_DOWNLOAD_SIZE / size_per_batch), 1
+        )
 
         return min(cache_batch_count_estimate, max_cache_batch_count)
 
+    def estimate_record_count_each_download(self, columns):
+        """Estimate the count of records with 20MB to download
+        """
+        record_size = self._estimate_record_size(columns)
+        max_count = int(MAX_DOWNLOAD_SIZE / record_size)
+        return max_count
+
+    def _estimate_record_size(self, columns):
+        """Estimate the average size of each record in the ODPS table.
+        """
+        sample_count = 10
```

Comment:
What’s this for?

Reply:
I have submitted another PR #1603 and the PR will be closed.